src/core/include/openvino/core/type.hpp

```diff
@@ -89,24 +90,44 @@ constexpr bool use_ov_dynamic_cast() {
     return std::is_base_of_v<ov::frontend::ConversionExtensionBase, T>;
 #endif
 }
+}  // namespace compile
+
+/// \brief Tests whether pointer can be static casted to a To*/shared_ptr<To>
+template <typename To, typename From>
+typename std::enable_if_t<
+    std::is_convertible_v<decltype(std::declval<From>().get_type_info().is_castable(To::get_type_info_static())), bool>,
+    bool>
+is_type(From* ptr) {
+    if constexpr (compile::use_ov_dynamic_cast<From>()) {
+        return ptr && ptr->get_type_info().is_castable(To::get_type_info_static());
+    } else {
+        static_assert(!std::is_volatile_v<To> && !std::is_volatile_v<From>, "is_type does not support volatile types");
+        return dynamic_cast<std::add_const_t<To>*>(ptr) != nullptr;
+    }
+}
 
-/// \brief Tests if value is a pointer/shared_ptr that can be statically cast to a
-/// Type*/shared_ptr<Type>
-template <typename Type, typename Value>
-typename std::enable_if<
-    std::is_convertible<decltype(std::declval<Value>()->get_type_info().is_castable(Type::get_type_info_static())),
-                        bool>::value,
-    bool>::type
-is_type(Value value) {
-    return value && value->get_type_info().is_castable(Type::get_type_info_static());
+/// \brief Tests whether pointer can be static casted to a To*/shared_ptr<To>
+template <typename To, typename From>
+typename std::enable_if_t<
+    std::is_convertible_v<
+        decltype(std::declval<std::shared_ptr<From>>()->get_type_info().is_castable(To::get_type_info_static())),
+        bool>,
+    bool>
+is_type(std::shared_ptr<From> ptr) {
```

Comment:
Is requried to copy the pointer for check only can be passed as reference?

Can inside be call just version for raw pointer?
```suggestion
is_type(std::shared_ptr<From> ptr) {
    return is_type<To, From>(ptr.get());
```
to avoid duplication?

Reply:
Applied.